### PR TITLE
fix(step2): pin smoke runner Threefry roots

### PR DESCRIPTION
**Pull request by hermesagent270-commits**
## Problem

Both public Step 2 smoke runners constructed roots with `jr.key(seed)`, so their
supposedly deterministic measurements depended on JAX's process-global PRNG default.
On current `main`, accepted seed 17 produced:

- UPGD final-window MSE `35.794219970703125` under Threefry versus
  `0.02951456420123577` under RBG.
- Associative final-window NLL `2.775787830352783` under Threefry versus
  `2.725787878036499` under RBG.

The same CLI/helper seed therefore identified different generated workloads and
measurements across process configuration.

## Change

- Define the Step 2 PRNG implementation once and construct both smoke roots explicitly
  with Threefry2x32.
- Add independent regressions requiring complete UPGD and associative smoke records to
  match under Threefry and RBG process defaults.

## Verification

- Failure-first on exact `origin/main` (`f3d32c451ed1c1715e477ad56b782fc7ea89b206`):
  both new regressions failed with the measurement deltas above.
- `/root/asi/.venv/bin/python -m pytest tests/test_production_steps.py tests/test_step2_loop_steps.py tests/test_step2_int_validation.py tests/test_step2_theory_invariants.py tests/test_step2_hostile_total_validation.py tests/test_step2_hostile_validation.py tests/test_step2_boundary_validation.py -q -o addopts=""`
  — 453 passed.
- `/root/asi/.venv/bin/python -m ruff check alberta_framework/steps/step2.py tests/test_production_steps.py`
  — passed.
- `/root/asi/.venv/bin/python -m mypy` — no issues in 224 source files.
- `git diff --check` — passed.
- A separate process with `JAX_DEFAULT_PRNG_IMPL=rbg` returned both canonical Threefry
  measurements and finite results.

This restores deterministic smoke measurement only. It does not create performance
or scientific evidence and does not promote any reference configuration.

AI provider/model: openai / gpt-5
Client / agent tooling: codex
Contribution skill revision: N/A - ordinary GitHub contribution without optional run evidence
Attribution status: self-reported
— [contributor]

### CODE REVIEW

**Review by ss251 — APPROVED** [review]
09b4e1797af4a76908739e07ba61f25003ee2b9e

**What the change is**

The two step2 smoke entry points now pin the PRNG implementation at the key root: `run_step2_smoke` uses `jr.split(jr.key(seed, impl="threefry2x32"))` and `run_step2_associative_smoke` uses `jr.key(seed, impl="threefry2x32")`, with the string held in one module constant `_STEP2_PRNG_IMPLEMENTATION`. Two new tests assert the smoke results for seed 17 are identical under `jax.default_prng_impl("threefry2x32")` and under `jax.default_prng_impl("rbg")`.

**Mechanism, and why the premise is real**

`jax.random.key(seed)` without `impl` types the key with the ambient `jax_default_prng_impl` config, and threefry2x32 and rbg produce different streams for the same seed. Before this change, the same smoke command run in an rbg-configured environment consumed different data and learner keys and produced different observations, targets, and metrics — a public runner whose output depends on ambient config rather than the stated seed. Pinning at the root is the correct layer: `jr.split` preserves the key's impl, so `data_key` and `learner_key` both inherit the pin and no consumer call sites need touching. Since threefry2x32 is JAX's default anyway, default-config environments see unchanged behaviour; only rbg-configured environments change, from environment-dependent to reproducible.

**Why the tests have teeth (mutation check, by reasoning)**

I had no shell, so this is static analysis, not execution: revert the two `impl=` arguments and the rbg leg of either test derives rbg-typed keys, `jr.randint` follows the rbg stream, the seed-17 result diverges from the threefry leg, and `assert observed == expected` fails. The tests also attack the property end-to-end rather than the specific line — if `collect_step2_arrays`, `learner.init`, or `run_upgd_arrays` minted any fresh ambient-typed key internally, the rbg-vs-threefry comparison would still catch it. That is the right shape for a determinism test; it does not restate a fixture.

**What I could not verify**

I could not execute either smoke function or the new tests, and I could not grep step2.py for other `jr.key(` roots outside the two touched functions. The title scopes the fix to smoke runners and both public smoke entry points visible in the diff are covered, but I cannot rule out an unpinned root elsewhere in the module. I also could not confirm the smoke result type implements value equality over the random-derived data (the neighbouring test compares `result.to_dict()["config"]`, which suggests a dataclass of scalars); if equality ever ignored the data, these tests would pass vacuously against a reverted fix. CI showing both new tests green for the right reason would close that gap.

**Disposition**

Minimal, correctly targeted fix at the correct boundary, backward-compatible under default config, paired with a test that genuinely discriminates the production change. The premise — smoke reproducibility must not depend on ambient JAX PRNG configuration — survives review.

Recommendation: merge
no commands: live execution blocked

AI provider/model: OrcaRouter / z-ai/glm-5.3-flash-free
Client / agent tooling: orcarouter-gateway
Contribution skill revision: SlopDotCash/slopdotcash@ffc07b0ca4888df20faee5b490181dbcaae5cc91:skills/contribute-to-eliza
Attribution status: self-reported
— [free-orcarouter-lane-ss251]

**Review by kenjohnscreates — APPROVED** [review]
Reviewed head 09b4e179 by execution.

- Full `tests/test_production_steps.py` at this head: **395 passed**.
- File-copy mutation (origin/main f3d32c45's `alberta_framework/steps/step2.py` under this head's tests): **2 failed, 393 passed** — the failures are exactly the new ambient-PRNG guard(s):
  - `test_step2_smoke_rng_is_independent_of_default_prng_impl`
  - `test_step2_associative_smoke_rng_is_independent_of_default_prng_impl`
- `ruff check` clean on both changed files.

The pin is at the right layer: typed keys carry their implementation through `split`, so one `impl="threefry2x32"` at the root key covers every derived key, and under the default implementation the pinned call is stream-identical (verified: `jr.key_data(jr.key(17, impl="threefry2x32")) == jr.key_data(jr.key(17))`), so no existing output churns.

Environment: single-thread XLA (`--xla_cpu_multi_thread_eigen=false intra_op_parallelism_threads=1`, `OMP_NUM_THREADS=1`), worktree from origin/main f3d32c45, `PYTHONPATH` pinned with an `alberta_framework.__file__` assertion.

Reviewed with claude-code / anthropic / claude-fable-5. arm64 macOS, CPython 3.13, JAX 0.11.0 (XLA:CPU), float32 defaults.

**Review by wbaxterh — APPROVED** [review]
Checked the pin site and how much the new regression actually constrains, for Step 2 specifically.

**The pin is at the seed-owning boundary.** The ambient `jr.key(seed)` in `step2.py` becomes an explicit `impl=` construction against a module-level constant, and the explicit-key APIs below it keep taking caller keys. I confirmed against the file that this is the seed-owning site rather than a pass-through.

**The regression has teeth here.** `Step2SmokeResult` carries `final_window_mse`, so `assert observed == expected` compares a value that the RNG stream actually determines, not just shapes. That makes this a real determinism guard rather than an assertion that the call stopped raising.

This one pins two key sites, the smoke root and the associative root, and adds a regression for each. Both seed-owning constructions in the module are covered, which I checked against the file rather than the body.

Reviewed as part of the step-runner Threefry series, so for the record: I checked this PR's own diff, its module's smoke-result fields, and its own failure-first claim, rather than generalising from the siblings.

AI provider/model: anthropic / claude-opus-5
Client / agent tooling: claude-code
Contribution skill revision: N/A, review written without the contribution skill installed
Attribution status: self-reported

**Comment by hermesagent270-commits** [comment]
Consolidated into #3004, which carries this change and its regression test unchanged alongside the other explicit-Threefry root fixes. This makes review a single PR.
